File: Python/isIsomorphic.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: isIsomorphic.py
@Function: 同构字符串
@Link: https://leetcode-cn.com/problems/isomorphic-strings/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-28
"""


class Solution:
    def isIsomorphic(self, s, t):
        x = {}
        y = {}
        for i in range(len(s)):
            if (s[i] in x and x[s[i]] != t[i]) or (
                    t[i] in y and y[t[i]] != s[i]):
                return False
            x[s[i]] = t[i]
            y[t[i]] = s[i]
        return True


# ----------------------------------
if __name__ == "__main__":
    s = "paper"
    t = "title"

    solution = Solution()
    check_isomorphic = solution.isIsomorphic(s, t)
    print(f"The solution of this problem is : {check_isomorphic}")

File: C++/isIsomorphic.cpp
/**
 * @File    : isIsomorphic.cpp
 * @Brief   : 同构字符串
 * @Link    : https://leetcode-cn.com/problems/isomorphic-strings/
 * @Author  : Wei Li
 * @Date    : 2021-07-28
*/

#include <iostream>
#include <string>
#include <vector>

class Solution
{
public:
    bool isIsomorphic(std::string s, std::string t)
    {
        std::vector<int> s_first_index(256, 0), t_first_index(256, 0);
        for (int i = 0; i < s.length(); ++i)
        {
            if (s_first_index[s[i]] != t_first_index[t[i]])
            {
                return false;
            }
            s_first_index[s[i]] = t_first_index[t[i]] = i + 1;
        }
        return true;
    }
};

// ----------------------------
int main(int argc, char **argv)
{
    std::string s = "paper", t = "title";

    Solution solution;
    bool check_isomorphic = solution.isIsomorphic(s, t);
    std::cout << "The solution of this problem is " << std::boolalpha << check_isomorphic << std::endl;

    return 0;
}
